awx/main/models/workflow.py

```diff
@@ -686,17 +698,59 @@ def _get_parent_field_name(self):
     def approve(self, request=None):
         self.status = 'successful'
         self.save()
+        self.send_approval_notification('approved')
         self.websocket_emit_status(self.status)
         schedule_task_manager()
         return reverse('api:workflow_approval_approve', kwargs={'pk': self.pk}, request=request)
 
     def deny(self, request=None):
         self.status = 'failed'
         self.save()
+        self.send_approval_notification('denied')
         self.websocket_emit_status(self.status)
         schedule_task_manager()
         return reverse('api:workflow_approval_deny', kwargs={'pk': self.pk}, request=request)
 
+    def signal_start(self, **kwargs):
+        can_start = super(WorkflowApproval, self).signal_start(**kwargs)
+        self.send_approval_notification('running')
+        return can_start
+
+    def send_approval_notification(self, status):
+        from awx.main.tasks import send_notifications  # avoid circular import
+        if self.workflow_job_template is None:
+            return
+        for nt in self.workflow_job_template.notification_templates["approvals"]:
+            try:
+                (notification_subject, notification_body) = self.build_notification_message(nt, status)
+            except AttributeError:
+                raise NotImplementedError("build_notification_message() does not exist" % status)
+
+            def send_it(local_nt=nt, local_subject=notification_subject, local_body=notification_body):
```

Comment:
This loop / function was copied from other places where the function needed to be redefined for each nt because the notification content could be different for each nt. As an aside, it's also worth noting that if don't define the function w/ kwargs, then the func would only get defined once due to late-binding - see https://github.com/ansible/awx/blob/devel/awx/main/models/notifications.py#L483-L488

I'm not 100% sure if in _this context_ `self.build_notification_message(nt, status)` will return a different message.  If I'm reading the PR correctly, though, it def. seems like they could. Here, for example:

![image](https://user-images.githubusercontent.com/4440360/64649978-e9c36380-d3d2-11e9-984a-4b91a475d913.png)

.. seems like the `workflow_url` would be unique for each node.

Reply:
In fact @beeankha, think it would be worth adding the comment that's here https://github.com/ansible/awx/blob/devel/awx/main/models/notifications.py#L483-L488 to this block so that we don't reintroduce https://github.com/ansible/awx/issues/4533

Reply:
Thanks @jbradberry and @jladdjr.  I've added that comment to this block of code.